File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        
        int maxP = 0;
        
        int minSoFar = prices[0];
        
        for(int i = 0 ; i < prices.length ; i++){
            minSoFar = Math.min(minSoFar,prices[i]);
            
            maxP = Math.max(maxP,prices[i]-minSoFar);
        }
        
        return maxP;
        
        
        
        
//         int buy = 0;
//         int sell = 1;
        
//         int maxP = 0;
        
//         while(sell<prices.length){
//             if(prices[sell]<prices[buy]){
//                 buy = sell;
//             }else{
//                 maxP = Math.max(maxP,prices[sell]-prices[buy]);
//             }
//             sell++;
//         }
        
//         return maxP;
    }
}

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution(object):
    def maxProfit(self, arr):
        """
        :type prices: List[int]
        :rtype: int
        """
        l,r = 0,1
        maxP=0
        
        while r<len(arr):
            if arr[l]<arr[r]:
                profit = arr[r]-arr[l]
                maxP = max(maxP,profit)
            else:
                l = r
            r+=1
            
            
        return maxP
        